File: sapphire/examples/kvstorejs/src/main/java/sapphire/appdemo/KeyValueStoreClient.java
package sapphire.appdemo;

import sapphire.app.SapphireObjectSpec;
import sapphire.common.SapphireObjectID;
import sapphire.kernel.server.KernelServerImpl;
import sapphire.oms.OMSServer;

import javax.xml.crypto.dsig.keyinfo.KeyValue;
import java.io.File;
import java.net.InetSocketAddress;
import java.nio.file.Files;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;
import java.util.List;

/**
 * Client class for testing {@link KeyValueStore_Stub}
 */
public class KeyValueStoreClient {
    public static void main(String[] args) throws Exception {
        OMSServer oms = getOMS(args[0], args[1]);

//        String spec = getSpec();
//        SapphireObjectID oid = oms.createSapphireObject(spec);
//        KeyValueStore_Stub store = (KeyValueStore_Stub)oms.acquireSapphireObjectStub(oid);
//        store.$__initializeGraal(SapphireObjectSpec.fromYaml(spec));

        KeyValueStore_Stub store = KeyValueStore_Stub.getStub(
                "KeyValueStore.yaml",
                args[0], args[1]);

        for (int i=0; i<30; ++i) {
            String key = "key_" + i;
            String val = "val_" + i;

            System.out.println(String.format("<client>: setting %s = %s", key, val));
            store.set(key, val);
            val = String.valueOf(store.get(key));
            System.out.println(String.format("<client>: got value %s with key %s", val, key));
        }
    }

    private static OMSServer getOMS(String omsIp, String omsPort) throws Exception {
        new KernelServerImpl(new InetSocketAddress("127.0.0.2", 0), new InetSocketAddress(omsIp, Integer.parseInt(omsPort)));
        Registry registry = LocateRegistry.getRegistry(omsIp, Integer.parseInt(omsPort));
        OMSServer omsserver = (OMSServer) registry.lookup("SapphireOMS");
        return omsserver;
    }

    private static String getSpec() throws Exception {
        ClassLoader classLoader = new KeyValueStoreClient().getClass().getClassLoader();
        File file = new File(classLoader.getResource("KeyValueStore.yaml").getFile());
        List<String> lines = Files.readAllLines(file.toPath());
        return String.join("\n", lines);
    }
}

File: sapphire/examples/kvstorejs/src/main/js/sapphire/appdemo/KeyValueStoreClient.js
/**
 * Client for testing {@link KeyValueStore}
 */

function main() {
    // get init configurations
    var system = Java.type("java.lang.System");
    var hostIP = system.getenv("HOST_IP");
    var hostPort = system.getenv("HOST_PORT");
    var omsIP = system.getenv("OMS_IP");
    var omsPort = system.getenv("OMS_PORT");

    // DM config file
    var ClientStub = Java.type("sapphire.appdemo.stubs.KeyValueStore_Stub");
    var kvs = ClientStub.getStub("KeyValueStore.yaml", omsIP , omsPort, hostIP, hostPort);

    for (i=0; i<3000; ++i) {

        // Sleep for a while so we can see the output on console during failover.
        // Note setTimeout does not work in graal vm so we use this while loop to sleep.
        var now = new Date().getTime();
        while ( new Date().getTime() < now + 1000 ) {}

        var key = "key_" + i;
        var val = "val_" + i;
        console.log(`client: setting ${key} = ${val}`);
        kvs.set(key, val);
        var v = kvs.get(key);
        console.log(`client: got value ${v} with key ${key}`);
    }
}
main();
